### ISSUE TITLE
[Feature] Folder uploads

### ISSUE THREAD

**Issue by reidbuzby** [microsoft/playwright#6854]
I am trying to upload a folder using playwright. I have been able to successfully use 
```javascript
page.on('filechooser', async fileChooser => {
  await fileChooser.setFiles('path/path_to_file.txt');
});
```
To upload _files_, however I want to upload a _folder_ instead. Something like:
```javascript
page.on('filechooser', async fileChooser => {
  await fileChooser.setFiles('path/path_to_folder/');
});
```
I have not found a way to do this after looking through the docs and searching Google, so I am guessing it is not currently supported. If I do try running that exact code above on a folder path I get the error:
```
fileChooser.setFiles: elementHandle.setInputFiles: EISDIR: illegal operation on a directory, read
```

I am hoping for a way to automate the process that would happen when clicking on `<input type=folder>`. Maybe the page could emit a different event 'folderchooser'.

**Comment by mxschmitt** [comment]
If you want to upload all the files from a directory, you can do the following:

```js
const fs = require('fs');
const path = require('path');

const testFolder = path.join(__dirname);

console.log(fs.readdirSync(testFolder).map(file => !fs.statSync(file).isDirectory() ? path.join(testFolder, file) : null).filter(Boolean));
```

This will read all the files and folders in the directory, maps over them to check if its a file and if not filters it out via the Boolean filter.

**Comment by pavelfeldman** [comment]
Let's collect user votes for this one!

**Comment by KapustinMaxim** [comment]
This feature will be very useful!

**Comment by maximkapusttin** [comment]
Add please

**Comment by robertyoung1993** [comment]
its very useful

**Comment by kentz007** [comment]
Let's collect user votes for this one!

**Comment by eliw00d** [comment]
> Let's collect user votes for this one!

What number are you looking to reach for user votes before it's considered?

**Comment by f-o-w-l** [comment]
This would be very helpful. Right now, when I pass multiple relative file paths to `setInputFiles`, their paths are not being passed to my onChange listener, and the `files` attribute returns `null`.

Example test:

Folder input component:
```javascript
  ...
  const onFolderChange = () => {
    const files = ref.current ? ref.current.files : null
    if (files && files.length > 0) {
      for (let i = 0; i < files.length; i += 1) {
        console.log(`name: "${files[i].name}"`)
        console.log(`path: "${files[i].path}"`)
        console.log(`webkitRelativePath: "${files[i].webkitRelativePath}"`)
      }
      // Doing stuff with file path...
    }
  }
  ...
  return (
    ...
    <input
      style={{ display: 'none' }}
      id="file"
      type="file"
      ref={ref}
      multiple
      onChange={onFolderChange}
    />
```
Test:
```javascript
test('folder can be opened and navigated', async () => {
  const electronApp = await electron.launch({ args: ['src/main.prod.js'] })
  const rendererWindow = await electronApp.firstWindow()

  rendererWindow.addListener('console', (msg) => {
    console.log('rendererWindow msg:', msg)
  })

  const folderInput = rendererWindow.locator('input#file')

  await folderInput.setInputFiles(['e2e/mocks/test_file_1.txt', 'e2e/mocks/test_file_2.txt'])

  console.log('files:', await folderInput.getAttribute('files'))

  await electronApp.close()
})
```

The files definitely exist, `ENOENT: no such file or folder` is not thrown. The file names are correct, but paths are not set.

Output:

```
rendererWindow msg: name: "test_file_1.txt"
rendererWindow msg: path: ""
rendererWindow msg: webkitRelativePath: ""
rendererWindow msg: name: "test_file_2.txt"
rendererWindow msg: path: ""
rendererWindow msg: webkitRelativePath: ""
files: null
```

Unfortunately this is breaking my e2e test as I perform verification on the file paths (checking for correct sub-folder structure within the selected folder) before doing anything with the files.

**Comment by araghuvarun** [comment]
#20939 is closed saying that it is duplicate. We've been waiting for this feature from long time. Could you pls prioritize this so that it would be useful not only to us but also many other users.

**Comment by gitnotgut** [comment]
I too have voted for this. Now that somany people are voting, is there any plan on adding this feature?

**Comment by masaplecas** [comment]
Any update on this?

**Comment by BStavniczki** [comment]
Any updates? It would be very useful.

**Comment by sv-automate** [comment]
Hi Team, we are trying to upload a folder in our scripts, do we have any alternate way either with third-party library or with script execution to upload the folder in playwright automation framework until we get this feature in the playwright itself, Thank you!

**Comment by pavelfeldman** [comment]
> Hi Team, we are trying to upload a folder in our scripts, do we have any alternate way either with third-party library or with script execution to upload the folder in playwright automation framework until we get this feature in the playwright itself, Thank you!

You probably can do that using scripts in Chromium, although I am not certain. It is best to paste your exact use case in this issue to make sure we cover it with the fix.

**Comment by haikyuu** [comment]
My use case is simply to test uploading a folder. The app handles file uploads and folder uploads differrently

**Comment by f-o-w-l** [comment]
I also have an app that behaves differently for files vs folders uploaded, probably could have mocked the <input> onChange but mocking the native File object was annoying so I gave up. Would love this feature!

**Comment by Jawwad1297Ahmed** [comment]
Hi 
Is there any workaround solution for this problem? 
please help

Thank you

**Comment by duty-beep** [comment]
My team needs this feature too

We are waiting for this to be merged (hopefully in next version) before migrating our project from Selenium to Playwright

**Comment by duty-beep** [comment]
Is it not planned to work on this in the near future?

**Comment by erezcor** [comment]
Hello! Any timeframe regarding this feature? @dgozman @mxschmitt @pavelfeldman @aslushnikov

**Comment by rubini881213** [comment]
Hello, Any new updates about folder uploads in Playwright ?

**Comment by mxschmitt** [comment]
This should be implemented [in Canary](https://playwright.dev/docs/canary-releases) and gets released as part of v1.45 soon. Feel free to beta-test or wait for v1.45 to happen.
